volume/stat.go

```diff
@@ -0,0 +1,138 @@
+	// %f - Raw mode in hex
+	// %n - File name
+	return []string{"find", path, "-mindepth", "1", "-exec", "stat", "-c", "%s %X %Y %Z %f %n", "{}", "+"}
+}
+
+// Keep as its own specialized function as it will be faster than munge.ToTime.
+func parseTime(t string) (time.Time, error) {
+	epoch, err := strconv.ParseInt(t, 10, 64)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return time.Unix(epoch, 0), nil
+}
+
+// StatParse parses a single line of the output of StatCmd into EntryAttributes and a name.
+func StatParse(line string) (plugin.EntryAttributes, string, error) {
+	var attr plugin.EntryAttributes
+	segments := strings.SplitN(line, " ", 6)
+	if len(segments) != 6 {
+		return attr, "", fmt.Errorf("Stat did not return 6 components: %v", line)
+	}
+
+	size, err := strconv.ParseUint(segments[0], 10, 64)
+	if err != nil {
+		return attr, "", err
+	}
+	attr.SetSize(size)
+
+	atime, err := parseTime(segments[1])
+	if err != nil {
+		return attr, "", err
+	}
+	attr.SetAtime(atime)
+
+	mtime, err := parseTime(segments[2])
+	if err != nil {
+		return attr, "", err
+	}
+	attr.SetMtime(mtime)
+
+	ctime, err := parseTime(segments[3])
+	if err != nil {
+		return attr, "", err
+	}
+	attr.SetCtime(ctime)
+
+	mode, err := munge.ToFileMode("0x" + segments[4])
+	if err != nil {
+		return attr, "", err
+	}
+	attr.SetMode(mode)
+
+	return attr, segments[5], nil
+}
+
+// Ensure the directory at newpath - and its parents - all exist. Return the contents of that
+// directory. The directory may not exist if we're parsing stat output for a basepath that's closer
+// to the root directory than where we searched because we want to preserve some of the hierarchy.
+func makedirs(attrs DirMap, newpath string) Dir {
+	// If it exists, return the children map. Base case would be newpath == "", which we create at
+	// the start of StatParseAll.
+	if newchildren, ok := attrs[newpath]; ok {
+		return newchildren
+	}
+
+	// Create the attributes map for the new path.
+	newchildren := make(Dir)
+	attrs[newpath] = newchildren
+
+	// Check if we need to create the parent, and get its attributes map.
+	parent, file := path.Split(newpath)
+	parent = strings.TrimSuffix(parent, "/")
+	parentchildren := makedirs(attrs, parent)
+
+	// Add attributes for the new path to the parent's attributes map. Then return the new map.
+	attr := plugin.EntryAttributes{}
+	attr.SetMode(os.ModeDir | 0550)
+	parentchildren[file] = attr
+	return newchildren
+}
+
+// StatParseAll an output stream that is the result of running StatCmd. Strips 'base' from the
+// file paths, and maps each directory to a map of files in that directory and their attr
+// (attributes).
+func StatParseAll(output io.Reader, base string) (DirMap, error) {
+	scanner := bufio.NewScanner(output)
+	// Create lookup table for directories to contents, and prepopulate the root entry because
+	// the mount point won't be included in the stat output.
+	attrs := DirMap{"": make(Dir)}
```

Comment:
I'd rename `attrs` to `dirMap` or `dirMp`, i.e. something that is more consistent with the return type.